text/0140-record-feature-flags-on-error.md

```diff
@@ -0,0 +1,47 @@
+- Start Date: 2024-09-04
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/140
+- RFC Status: draft
+
+# Summary
+
+As feature flags are evaluated in a customer's application we will collect those flags and their results, store them in-memory, and push the values to Sentry when an error occurs. The flags can be used to further understand what went wrong during the session.
+
+# Motivation
+
+To enable users to debug application errors with a complete picture of their application's state.
+
+# Options Considered
+
+We will collect flag evaluations and hold them in-memory. On error event the flags will be placed into the `contexts` object on the `event` and sent to Sentry.
+
+## Transport
+
+The flags will be represented by this data structure during transport. The `flag` key is the name of the flag and the `result` key is the evaluation result returned by the customer's application.
+
+```json
+{
+  "contexts": {
+    "flags": {
+      "values": [
+        { "flag": "abc", "result": true },
+        { "flag": "def", "result": false }
+      ]
+    }
+  }
+}
+```
+
+## Public Interface
+
+The SDK will expose one new public method `set_flag`. The method accepts the arguments `flag` (of type string) and `result` (of type boolean). Similar to `set_tag` or `set_user`, the `set_flag` method stores a flag, result pair on the isolation scope. On error, the isolation scope's flags are serialized and appended to the event body as described in the previous section.
+
+Stateless, multi-tenanted applications, such as web servers, must isolate flag evaluations per request.
+
+## Bounding Memory Usage and Transport Size
+
+The number of flag evaluations must be capped to some fixed capacity (e.g. 100). When the capacity is reached the least recently evaluated flag should be dropped from the set. Duplicate flag evaluations may update their existing entry rather than insert a new one. In most contexts, it will make sense to update the existing entry. An LRU cache is recommended for these cases. If repeat flag evaluations are desired then a ring buffer may be used.
```

Comment:
I'm wondering if this is strictly necessary? Do we have backend requirements to keep the number of flags capped to e.g. 100? SDK-side, we have normalization and trimming logic to keep event sizes under control, as well as a [maxbreadth attribute](https://github.com/getsentry/sentry-javascript/blob/1e9a1a3c80b10d2aca19b640666af9134d3bf40a/packages/types/src/options.ts#L136-L156) to control how many keys an object can contain. 

Also, duplicated flags should be fairly easy to avoid (at least in JS) via `setFlags` but let's not forget that users could always overwrite the entire content via `setContext('flags', ...)`.

Reply:
Malformed code which inserts unique flags in a loop would present as a memory leak.  Whether or not this is reason for us to put in a defense mechanism I suppose is a matter of Sentry's opinion.